### PR TITLE
add mm_fp4 use cudnn backend

### PR DESCRIPTION
**Pull request by ttyio**
support a/b input type e2m1, block quant type e4m3 with block size 16 output bfloat16 and fp16.

## 📌 Description

init add mm_fp4 use cudnn backend

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

### CODE REVIEW

File: flashinfer/gemm.py [hunk 1]
```diff
@@ -872,6 +986,187 @@ def _cudnn_gemm_fp8(
...
+    --------
+    >>> import torch
+    >>> from flashinfer import fp4_quantize, mm_fp4
+    >>> a = torch.randn([48, 128], device="cuda", dtype=torch.bfloat16)
+    >>> b = torch.randn([256, 128], device="cuda", dtype=torch.bfloat16).transpose(-2, -1)
+    >>> a_global_sf = (448 * 6) / a.float().abs().nan_to_num().max()
+    >>> b_global_sf = (448 * 6) / b.float().abs().nan_to_num().max()
+    >>> a_fp4, a_sf = fp4_quantize(a, a_global_sf, 16, False, True)
+    >>> b_fp4, b_sf = fp4_quantize(b, b_global_sf, 16, False, True)
+    >>> out = mm_fp4(a_fp4, b_fp4, a_sf, b_sf, 1.0/(a_global_sf * b_global_sf), torch.bfloat16, None)
```
[the remark this answers is not in the document]
Comment (ttyio) [reply flashinfer/gemm.py:1151]: leave this unchange to make the sample code more short?

File: flashinfer/gemm.py [hunk 2]
```diff
@@ -872,6 +986,187 @@ def _cudnn_gemm_fp8(
...
+
+    # each packed uint8 contains 2 fp4 elements
+    real_shape[-2 if is_column_major else -1] *= 2
+    if is_column_major:
+        real_stride[-1] *= 2
+        for i in range(len(real_stride) - 2):
+            real_stride[i] *= 2
+    else:
+        for i in range(len(real_stride) - 1):
+            real_stride[i] *= 2
```
[the remark this answers is not in the document]
Comment (ttyio) [reply flashinfer/gemm.py:1064]: the minor dim always has stride 1, and other dims has stride multiply by 2, so code in the PR is correct?

File: flashinfer/gemm.py [hunk 3]
```diff
@@ -872,6 +986,187 @@ def _cudnn_gemm_fp8(
...
+    torch.Size([48, 256])
+    """
+    # pre-check the input tensor, block scale tensor and alpha tensor
+    if a.ndim != 2 or b.ndim != 2:
+        raise ValueError(f"mm_fp4 accepts 2d tensors, got {a.shape} and {b.shape}")
+    if a.shape[1] != b.shape[0]:
+        raise ValueError(
+            f"K dimension mismatch in mm_fp4. got a.shape[1] = {a.shape[1]}, b.shape[0] = {b.shape[0]}"
+        )
+    if a.dtype not in {torch.uint8, torch.float4_e2m1fn_x2} or b.dtype not in {
```
Comment (yzh119) [inline flashinfer/gemm.py]: Please also handle the logic of early torch versions (e.g. torch 2.7.1 stable do not support `torch.float4_e2m1fn_x2`).

File: flashinfer/gemm.py [hunk 4]
```diff
@@ -872,6 +986,187 @@ def _cudnn_gemm_fp8(
...
+    batch = real_a_shape[0]
+    expanded_a_descale_shape, expanded_a_descale_stride = (
+        _expand_block_scale_tensor_shape(a_descale, batch)
+    )
+    expanded_b_descale_shape, expanded_b_descale_stride = (
+        _expand_block_scale_tensor_shape(b_descale, batch)
+    )
+
+    # build the fp4 cudnn graph
+    FP4_BLOCK_SIZE = 16
```
[the remark this answers is not in the document]
Comment (ttyio) [reply flashinfer/gemm.py]: this `FP4_BLOCK_SIZE` only lived in this `mm_fp4` function, no need to add to module-level and exposed.

File: flashinfer/gemm.py [hunk 5]
```diff
@@ -872,6 +986,187 @@ def _cudnn_gemm_fp8(
...
+    a_descale: torch.Tensor
+        Block scale tensor for A, shape (m, k // block_size), float8_e4m3fn or uint8.
+
+    b_descale: torch.Tensor
+        Block scale tensor for B, shape (k, n // block_size), float8_e4m3fn or uint8.
+
+    alpha: torch.Tensor
+        Global scale tensor, float scalar.
+
+    dtype: torch.dtype
```
Comment (yzh119) [inline flashinfer/gemm.py]:
```suggestion
    out_dtype: torch.dtype
```
sounds more clear to me.

File: flashinfer/gemm.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (yzh119) [reply flashinfer/gemm.py]: Please also document it clearly in the documentation.

File: flashinfer/gemm.py [hunk 6]
```diff
@@ -872,6 +986,187 @@ def _cudnn_gemm_fp8(
...
+    Parameters
+    ----------
+    a: torch.Tensor
+        Input tensor, shape (m, k), fp4 e2m1fn_x2 or uint8.
+
+    b: torch.Tensor
+        Mat2 tensor, shape (k, n), should be column major, fp4 e2m1fn_x2 or uint8.
+
+    a_descale: torch.Tensor
+        Block scale tensor for A, shape (m, k // block_size), float8_e4m3fn or uint8.
```
Comment (yzh119) [inline flashinfer/gemm.py:1119]: The `block_size` is not mentioned in the docstring and might confuse people.

We can still add a block_size argument (and when it's not 16, we raise runtime errors).

Comment (yzh119) [reply flashinfer/gemm.py:1119]: Examples include https://github.com/flashinfer-ai/flashinfer/blob/e885dd0c985b7f90e668949e26aad52c02c7e074/flashinfer/gemm.py#L964

File: tests/test_mm_fp4.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        a.cuda(),
+        a_global_sf.cuda(),
+        sf_vec_size,
+        sf_use_ue8m0=False,
+        is_sf_swizzled_layout=True,
+    )
+    return a_fp4, a_sf, a_global_sf
+
+
+@pytest.mark.parametrize("m", [48, 128])
```
Comment (yzh119) [inline tests/test_mm_fp4.py]: Can we add more problem shapes for unittests?

File: flashinfer/gemm.py [hunk 8]
```diff
@@ -752,6 +762,115 @@ def _validate_fp8_output_dtype(dtype: torch.dtype):
...
+    b_descale_shape,
+    b_descale_stride,
+    ab_type,
+    scale_type,
+    o_type,
+    block_size,
+):
+    _check_cudnn_availability()
+
+    with cudnn.graph(_get_cudnn_handle()) as (graph, _):
```
Comment (elfiegg) [inline flashinfer/gemm.py]: May we please add something like
```
stream = torch.cuda.current_stream().cuda_stream
cudnn.set_stream(handle=cudnn_handle, stream=stream)
```

to ensure CUDA graph can be captured on the right device?

**Comment by elfiegg** [comment]
Actually I just tried setting the stream of cudnn but still got
```
/usr/local/lib/python3.12/dist-packages/torch/cuda/graphs.py:84: UserWarning: The CUDA Graph is empty. This usually means that the graph was attempted to be captured on wrong device or stream. (Triggered internally at /opt/pytorch/pytorch/aten/src/ATen/cuda/CUDAGraph.cpp:175.)
```
cc @Anerudhan for suggestions

**Comment by yzh119** [comment]
@elfiegg can you put your script for capturing the kernel in cudagraph mode?

**Comment by elfiegg** [comment]
Two suggestions from the local test from my end:
1. Ensure cached graph is built on the same stream as input, and
2. Ensure graph is executed on the same stream as input. 
Both via adding:
```
    stream = torch.cuda.current_stream(a_device).cuda_stream
    cudnn.set_stream(handle=_get_cudnn_handle(), stream=stream)

```

For repro @yzh119 :
Benchmark script: 
https://gitlab-master.nvidia.com/elfieg/benchmarks/-/blob/cudnn-fp4/benchmark_fp4_gemm.py?ref_type=heads
Container: `gitlab-master.nvidia.com:5005/dl/dgx/sglang:main-py3.28056744-base`

**Comment by Anerudhan** [comment]
Sorry was out.  Please look at https://github.com/flashinfer-ai/flashinfer/pull/1283/files#diff-c41e577ab076c5789a1371ced63bed17e95b83f69b2376f589a38d464c00f8acR231

**Comment by ttyio** [comment]
Thank you all for the help on cuda graph issue, with the latest fix to always update the cuda stream when retrieve the cudnn handle. I am able to run @elfiegg 's attached benchmark

```
fp4_gemm_benchmark:
    batch_size  baseline cutlass fp4  cudnn fp4
0          1.0              0.011813   0.011807
1          2.0              0.011822   0.011818
2          4.0              0.011817   0.011818
3          8.0              0.011837   0.011834
4         16.0              0.011910   0.011908
5         32.0              0.012041   0.012039
......
```

**Comment by Anerudhan** [comment]
> Thanks @ttyio for creating this PR and thanks @elfiegg @Anerudhan for your input.
> 
> I tried this PR on different environments and it seems dependency and version control is still a headache:
> 
> * the fp4 gemm feature is only available for latest cudnn version, and we we have to upgrade three dependencies:
>   
>   * cudnn installation in system (9.11), which can not be managed by pip
>   * nvidia-cudnn-frontend package (1.13)
>   * nvidia-cudnn-cu12 package (9.11.0.98)
> * using latest `nvidia-cudnn-cu12` will be conflict with the dependency of torch earlier than 2.8.*
> 
> Considering this, I encourage adding some runtime checks:
> 
> 1. using cudnn fp4 gemm requires latest version of cudnn + cudnn-frontend + cudnn-cu12 package and if dependencies version doesn't meet, let's raise a runtime error.
> 2. if the cudnn version do not meet the requirement of fp4 gemm, the other components of flashinfer should still work.

With minor version compatibility, we can run with older version of torch, right? 
To reiterate, I vision. something like torch2.7 + cuda12.8 + cudnn 9.11

We can just error out with "you need cudnn-9.11" to get perf and allow FI to fallback to a different backend. 

Also, note that fp4 is not supported anyway in torch 2.8 and before ?

**Comment by ttyio** [comment]
> * torch earlier than 2.8

@yzh119 @Anerudhan I added cudnn fe >= 1.13.* and cudnn backend >= 91002 check, could you help to review if this make sense? thanks!

**Comment by yzh119** [comment]
> Also, note that fp4 is not supported anyway in torch 2.8 and before ?

Yes but user can still use uint8 containers to use other fp4 kernels.

**Review by yzh119 — APPROVED** [review]
Overall LGTM, and thank all of you for the effort.

cc @Anerudhan for another look on @ttyio 's most recent commit

**Comment by ttyio** [comment]
@yzh119 could you help me to merge the PR since there is no other issue per slack channel discussion, thank you!
